### ISSUE TITLE
@Blocking and @Incoming should support return types of CompletionStage and Uni

### ISSUE THREAD

**Issue by sorin0187** [quarkusio/quarkus#11801]
I tried creating a method that persists an entity upon receiving a Kafka message. 

```
@Incoming("input-topic")
@Blocking
public void processPrinterEvent(final Message<PrinterMessage> message) { 
 // create and persist entity
}
```

When building I get the following exception:
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.5.1.Final:build (default) on project seq-pmd-backend: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.smallrye.reactivemessaging.deployment.SmallRyeReactiveMessagingProcessor#build threw an exception: javax.enterprise.inject.spi.DefinitionException: Invalid method annotated with @Incoming: com.bmw.otdlip.sepia.pmd.sequencestream.integration.kafka.boundary.PrinterTopicConsumer#processPrinterEvent - The signature is not supported as it requires 'blocking' acknowledgment, return a CompletionStage<Message<?> instead.
[ERROR]         at io.smallrye.reactive.messaging.MediatorConfigurationSupport.getIncomingError(MediatorConfigurationSupport.java:548)
[ERROR]         at io.smallrye.reactive.messaging.MediatorConfigurationSupport.validateSubscriber(MediatorConfigurationSupport.java:172)
[ERROR]         at io.smallrye.reactive.messaging.MediatorConfigurationSupport.validate(MediatorConfigurationSupport.java:84)
[ERROR]         at io.quarkus.smallrye.reactivemessaging.deployment.QuarkusMediatorConfigurationUtil.create(QuarkusMediatorConfigurationUtil.java:80)
[ERROR]         at io.quarkus.smallrye.reactivemessaging.deployment.SmallRyeReactiveMessagingProcessor.build(SmallRyeReactiveMessagingProcessor.java:315)
[ERROR]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:932)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR]         at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
[ERROR]         at java.lang.Thread.run(Thread.java:748)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

If I try to return a CompletionStage, as suggested by the exception, I get a different validation failing:
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.5.1.Final:build (default) on project seq-pmd-backend: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.smallrye.reactivemessaging.deployment.SmallRyeReactiveMessagingProcessor#build threw an exception: javax.enterprise.inject.spi.DefinitionException: Invalid method annotated with @Blocking: com.bmw.otdlip.sepia.pmd.sequencestream.integration.kafka.boundary.PrinterTopicConsumer#processPrinterEvent - The @Blocking annotation is only supported for methods returning an individual Message or a payload
[ERROR]         at io.smallrye.reactive.messaging.MediatorConfigurationSupport.getBlockingError(MediatorConfigurationSupport.java:540)
[ERROR]         at io.smallrye.reactive.messaging.MediatorConfigurationSupport.validateBlocking(MediatorConfigurationSupport.java:529)
[ERROR]         at io.quarkus.smallrye.reactivemessaging.deployment.QuarkusMediatorConfigurationUtil.create(QuarkusMediatorConfigurationUtil.java:122)
[ERROR]         at io.quarkus.smallrye.reactivemessaging.deployment.SmallRyeReactiveMessagingProcessor.build(SmallRyeReactiveMessagingProcessor.java:315)
[ERROR]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:932)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR]         at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
[ERROR]         at java.lang.Thread.run(Thread.java:748)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:479)
[ERROR] -> [Help 1]
```

It looks like it's not allowed to have both `@Blocking` and `@Incoming` annotations on a method that receives a Message<>.
The reason I need this is because the topic I'm consuming messages from is a compacted topic, and in case of a tombstone message (null payload) I want to delete the entity based on the message key.

**Comment by kenfinnigan** [comment]
What was the code using the `CompletionStage` you tried?

I think you hit https://github.com/smallrye/smallrye-reactive-messaging/blob/master/smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/MediatorConfigurationSupport.java#L169 and maybe it's a missed update that we've created error loop that needs to be resolved. @cescoffier might know if that error should be adjusted.

For your example, this should work:

```
@Incoming("input-topic")
@Blocking
public void processPrinterEvent(final PrinterMessage printerMessage) { 
 // create and persist entity
}
```

**Comment by cescoffier** [comment]
We should be able to support this. Basically, we could off-load the call to the worker. Something to fix in reactive messaging.

**Comment by kenfinnigan** [comment]
I've created https://github.com/smallrye/smallrye-reactive-messaging/issues/718 to implement the feature for returning `Uni` and `CompletionStage`

**Comment by sorin0187** [comment]
> The first error message is present because `void myMethod(Message<PrinterMessage> message)` is deprecated, it's not anything to do with `@Blocking`. If you remove `@Blocking` you should get the same error message.
> 
> The use of `CompletionStage` return type with `@Blocking` is not allowed, which is why you got the second error. But I think the second error could use some clarification for this particular case.
> 
> For your example, this should work:
> 
> ```
> @Incoming("input-topic")
> @Blocking
> public void processPrinterEvent(final PrinterMessage printerMessage) { 
>  // create and persist entity
> }
> ```

You're right. Your example does work, however it doesn't solve my issue, since in the case of a tombstone message the payload would be null, and I need the key to know which entity to remove from the DB.

I solved this by running the blocking operation on a different thread manually (instead of using the @Blocking annotation):
`   
    @Incoming("available-printer-in")
    public CompletionStage<Void> processPrinterEvent(final Message<PrinterMessage> message) {
        return CompletableFuture.runAsync(() -> processMessage(message)).thenRun(message::ack);
    }
`

**Comment by cescoffier** [comment]
The next version provides a better abstraction to handle `null` payload. You would be able to get a `Record<K, V>`, with both K and V nullable.

**Comment by jigneshmpatel** [comment]
We are using quarkus 2.8.0 and using blocking threads, I have implemented above suggested solution.

    public CompletionStage<Void> processEvent(Message<String> incomingjsonreceive){
         return CompletableFuture.runAsync(()-> processmessage(incomingjsonreceive.getPayload())).thenRun(incomingjsonreceive::ack);
    }

It works fine for the few messages, but if we go beyond 50+ messages. We are seeing following error for the consumer-group



Preparing Rebalance |   | A reassignment of partitions is required, members have been asked to stop consuming

Empty |   | Consumer group exists, but does not have any members

why the consumer group tries to rebalance is not understandable if there are more messages.
 
Dead |   | Consumer group does not have any members and it's metadata has been removed
 
Unknown |   | Group state is not known
